## README

This repository contains scripts for the following manuscript: https://www.biorxiv.org/content/10.1101/2023.09.14.557820v1

All scripts are hosted under **'SCRIPTS'** folder.

Input data that are required to run the scripts are hosted under **'IN_DATA'** folder.

Output data that are generated by the scripts are hosted under **'OUT_DATA'** folder.

Due to file size limit in github, some of the input files are too large to upload. We provide download links for these data below. We also list R package requirements with their versions and provide vignettes that provide explanations and instructions to run the scripts.

**Download Links:**

Multiz alignment files: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/multiz30way/maf/

<br>
<br>

### ***Package Requirements***

**R version: 4.1.1**

**R packages:**

DESeq2_1.34.0

edgeR_3.36.0       

limma_3.50.3

scater_1.22.0    

scran_1.22.1

scuttle_1.4.0   

SingleCellExperiment_1.16.0  

SummarizedExperiment_1.24.0   

MatrixGenerics_1.6.0     

matrixStats_0.61.0           

Matrix.utils_0.9.8     

patchwork_1.1.1              

org.Hs.eg.db_3.14.0    

rrvgo_1.6.0           

rGREAT_1.26.0       

gridExtra_2.3                    

tgutil_0.1.13           

BSgenome.Hsapiens.UCSC.hg38_1.4.4

BSgenome_1.62.0     

rtracklayer_1.54.0    

EnsDb.Hsapiens.v86_2.99.0   

ensembldb_2.18.4           

AnnotationFilter_1.18.0     

GenomicFeatures_1.46.1           

AnnotationDbi_1.56.1    

Biobase_2.54.0         

Matrix_1.5-3   

Signac_1.10.0      

GeneOverlap_1.30.0     

motifmatchr_1.16.0   

JASPAR2020_0.99.10   

TFBSTools_1.32.0    

GenomicRanges_1.46.0      

rio_0.5.29                     

data.table_1.14.2     

forcats_1.0.0        

stringr_1.5.0   

purrr_1.0.1        

tidyr_1.3.0    

tibble_3.1.8       

tidyverse_1.3.2      

plyr_1.8.6            

VennDiagram_1.7.3    

futile.logger_1.4.3    

msa_1.26.0       

phangorn_2.10.0       

seqinr_4.2-23   

readr_2.1.4          

pheatmap_1.0.12   

reshape2_1.4.4         

SeuratObject_4.1.3     

Seurat_4.3.0             

ggpubr_0.4.0       

ggplot2_3.4.2            

Biostrings_2.62.0     

GenomeInfoDb_1.35.15    

XVector_0.34.0          

IRanges_2.28.0         

S4Vectors_0.32.0    

BiocGenerics_0.40.0      

dplyr_1.1.0          

ape_5.6-2                  

rphast_1.6.11

<br>
<br>
  
### **Ancestral Substitutions**

The following scripts use the phylogenetic tree and the UCSC multiz alignment to construct ancestral states for each peak region. Then all DNA substitutions are catalogued based on when they occurred (e.g human lineage, hominin lineage) and peaks that accumulated significantly more DNA substitutions in one lineage than others are categorized accordingly.

**Step 1: Extract peak regions from the UCSC multiz alignment.**
Multiz alignment files are very large and the peak regions only cover a fraction of the genome. To make downstream analysis faster, this tool extracts and saves multiz alignment at the peak regions. At this step we also keep only the species with good genome quality.

```
#chr: Only one chromosome per run

#keepspecies_fn: File that has species to be extracted from multiz alignment

#msa_fn: Multiz alignment for the given chromosome

#peak_fn: scATAC-seq peak file

#nthreads: Number of threads for parallelization

#outdir: Folder to output all data generated by this script

Rscript SCRIPTS/Extract_GREs_From_Alignment.R \
			chr='chr20'\
			keepspecies_fn='IN_DATA/keepspecies.txt' \
			msa_fn='IN_DATA/chr20.maf' \
			peak_fn='IN_DATA/merged_lifted_human_sorted.bed' \
			nthreads=23 \
			outdir='OUT_DATA/Extracted_CRE_Alignments' \
```

**Step 2: Find ancestral substitutions.**
This script operates on the extracted multiz alignment at the peak regions and the phylogenetic tree to identify the substitutions that occurred on the five lineages examined in the paper (Human, Hominin, African Great Ape, Great Ape and Ape).

```
#submodel: Nucleotide substitution model

#treetop_fn: Tree topology file

#pks_fn: scATAC-seq peak file

#outdir: Folder to output all data generated by this script

#nthreads: Number of threads for parallelization

#chrs: Chromosomes

Rscript SCRIPTS/Extract_Ancestral_Substitutions.R \
			submodel='HKY' \
			treetop_fn='IN_DATA/10ktrees.nunn-lab.org.tree.newick' \
			pks_fn='IN_DATA/merged_lifted_human_sorted.bed' \
			outdir='OUT_DATA/Substitutions_Adult' \
			nthreads=23 \
			chrs='chr21,chr22' \
```














