## README

This repository contains scripts for the following manuscript: https://www.biorxiv.org/content/10.1101/2023.09.14.557820v1

All scripts are hosted under **'SCRIPTS'** folder.

Input data that are required to run the scripts are hosted under **'IN_DATA'** folder.

Output data that are generated by the scripts are hosted under **'OUT_DATA'** folder.

Due to file size limit in github, some of the input files are too large to upload. We provide download links for these data below. We also list R package requirements with their versions and provide vignettes that provide explanations and instructions to run the scripts.

**Download Links:**

Multiz alignment files: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/multiz30way/maf/

Other files (please download to IN_DATA folder): https://cloud.biohpc.swmed.edu/index.php/s/Y28GwDaBGW2XGf8

<br>
<br>

### ***Package Requirements***

**R version: 4.1.1**

**R packages:**

DESeq2_1.34.0

edgeR_3.36.0       

limma_3.50.3

scater_1.22.0    

scran_1.22.1

scuttle_1.4.0   

SingleCellExperiment_1.16.0  

SummarizedExperiment_1.24.0   

MatrixGenerics_1.6.0     

matrixStats_0.61.0           

Matrix.utils_0.9.8     

patchwork_1.1.1              

org.Hs.eg.db_3.14.0    

rrvgo_1.6.0           

rGREAT_1.26.0       

gridExtra_2.3                    

tgutil_0.1.13           

BSgenome.Hsapiens.UCSC.hg38_1.4.4

BSgenome_1.62.0     

rtracklayer_1.54.0    

EnsDb.Hsapiens.v86_2.99.0   

ensembldb_2.18.4           

AnnotationFilter_1.18.0     

GenomicFeatures_1.46.1           

AnnotationDbi_1.56.1    

Biobase_2.54.0         

Matrix_1.5-3   

Signac_1.10.0      

GeneOverlap_1.30.0     

motifmatchr_1.16.0   

JASPAR2020_0.99.10   

TFBSTools_1.32.0    

GenomicRanges_1.46.0      

rio_0.5.29                     

data.table_1.14.2     

forcats_1.0.0        

stringr_1.5.0   

purrr_1.0.1        

tidyr_1.3.0    

tibble_3.1.8       

tidyverse_1.3.2      

plyr_1.8.6            

VennDiagram_1.7.3    

futile.logger_1.4.3    

msa_1.26.0       

phangorn_2.10.0       

seqinr_4.2-23   

readr_2.1.4          

pheatmap_1.0.12   

reshape2_1.4.4         

SeuratObject_4.1.3     

Seurat_4.3.0             

ggpubr_0.4.0       

ggplot2_3.4.2            

Biostrings_2.62.0     

GenomeInfoDb_1.35.15    

XVector_0.34.0          

IRanges_2.28.0         

S4Vectors_0.32.0    

BiocGenerics_0.40.0      

dplyr_1.1.0          

ape_5.6-2                  

rphast_1.6.11

<br>
<br>
  
### **Ancestral Substitutions**

The following scripts use the phylogenetic tree and the UCSC multiz alignment to construct ancestral states for each peak region. Then all DNA substitutions are catalogued based on when they occurred (e.g human lineage, hominin lineage) and peaks that accumulated significantly more DNA substitutions in one lineage than others are categorized accordingly.

**Step 1: Extract peak regions from the UCSC multiz alignment.**
Multiz alignment files are very large and the peak regions only cover a fraction of the genome. To make downstream analysis faster, this tool extracts and saves multiz alignment at the peak regions. At this step we also keep only the species with good genome quality.

```
#chr: Only one chromosome per run

#keepspecies_fn: File that has species to be extracted from multiz alignment

#msa_fn: Multiz alignment for the given chromosome

#peak_fn: scATAC-seq peak file

#nthreads: Number of threads for parallelization

#outdir: Folder to output all data generated by this script

Rscript SCRIPTS/Extract_GREs_From_Alignment.R \
			chr='chr20'\
			keepspecies_fn='IN_DATA/keepspecies.txt' \
			msa_fn='IN_DATA/chr20.maf' \
			peak_fn='IN_DATA/merged_lifted_human_sorted.bed' \
			nthreads=23 \
			outdir='OUT_DATA/Extracted_CRE_Alignments' \
```

**Step 2: Find ancestral substitutions.**
This script operates on the extracted multiz alignment at the peak regions and the phylogenetic tree to identify the substitutions that occurred on the five lineages examined in the paper (Human, Hominin, African Great Ape, Great Ape and Ape).

```
#submodel: Nucleotide substitution model

#treetop_fn: Tree topology file

#pks_fn: scATAC-seq peak file

#outdir: Folder to output all data generated by this script

#nthreads: Number of threads for parallelization

#chrs: Chromosomes

Rscript SCRIPTS/Extract_Ancestral_Substitutions.R \
			submodel='HKY' \
			treetop_fn='IN_DATA/10ktrees.nunn-lab.org.tree.newick' \
			pks_fn='IN_DATA/merged_lifted_human_sorted.bed' \
			outdir='OUT_DATA/Substitutions_Adult' \
			nthreads=23 \
			chrs='chr21,chr22' \
```

**Step 3: Assign GREs to evolutionary groups.**
This script uses the raw substitution calls at each lineage from Step 2, plots basic statistics and performs statistics to identify peaks that have accumulated more DNA substitutions in each lineage compared to the other lineages. 

```

#subsdir: Folder that contains all substitutions

#outdir: Folder to output all data generated by this script

#plt_pref: Prefix to add to all plot names

#sd_cutoff: Standard deviation cutoff (e. g. 1 sd more divergent than the rest of the GREs within the node)

#fdr_cutoff=0.05: FDR cutoff for significant more divergence of a GRE in one node than the other nodes

#fc_cutoff=1.5: Fold change cutoff for more divergence of a GRE in one node than the other nodes

Rscript SCRIPTS/Assign_To_Evogroups.R \
			subsdir='OUT_DATA/Substitutions_Adult' \
			outdir='OUT_DATA/Evo_groups' \
			plt_pref='Adult' \
			sd_cutoff=1 \
			fdr_cutoff=0.05 \
			fc_cutoff=1.5 \
```

**Step 4: Cell type enrichment of evolutionary groups.**
This is a short script that runs Fisher's exact test between evolutionarily divergent peaks identified in Step 3 and the peaks that are more accessible per cell type.

```
#evo_group_fn: Combined list of GREs categorized by their evolutionary divergence. The script Assign_To_Evogroups provides this ouput as 'CREs_SignTested_ConsAdded.RDS'. Since this vignette only runs chromosomes 20-22, here we will feed the end result of the entire dataset

#ctmarks_fn: Celltype markers file in RDS format. Load the example file to R to see its format.

#bcg_size: Total number of peaks in the dataset. Used as a background in the enrichment analysis.

#plt_pref: Prefix to add to all plot names

Rscript SCRIPTS/Evogroups_Celltype_Enrichment.R \
			evo_group_fn='OUT_DATA/AdultCREs_SignTested_SD_1_FC_ConsAdded.RDS' \
			ctmarks_fn='IN_DATA/PseudoBulk_DARs_MAJOR_MARKERS.RDS' \
			bcg_size=364657 \
			plt_pref='Adult' \
```

<br>

### **TFBS Evolution**

The following scripts use the ancestral states constructed in the previous section and generate a peak-motif matrix for every lineage. It then uses this peak-motif matrix to identify events where a motif occurrence is gained or lost. For each motif, gain/loss ratio of is then used as a readout for the expansion/depletion of the TFBS in each lineage.
<br>
<br>

**Step 1: Create a list of TFBS gains and losses per lineage for all JASPAR motifs.**
This script calculates all motif gain and loss events for each lineage and saves the raw results for further analysis.

```
#submodel: Nucleotide substitution model

#pks_fn: scATAC-seq peak file

#treetop_fn: Tree topology file

#outdir: Folder to output all data generated by this script

#nthreads: Number of threads for parallelization

#chr: Chromosome

#extracted_align: Subset of multiz alignment

Rscript SCRIPTS/TFBS_Evolution.R \
			submodel='HKY' \
			pks_fn='IN_DATA/merged_lifted_human_sorted.bed' \
			treetop_fn='IN_DATA/10ktrees.nunn-lab.org.tree.newick' \
			outdir='OUT_DATA/TFBS_Evo_Adult' \
			chr="chr22" \
			extracted_align="OUT_DATA/Extracted_CRE_Alignments/chr22.RDS" \
			nthreads=23
```

**Step 2: Analyze TFBS gain/loss changes across lineages.**
This script identifies significantly expanded or depleted motifs per lineage.

```

#new_or_ms: Should the TFBS gain/loss file be the one in the manuscript (MS) or generated new using the previous command (NEW)? If MS is chosen, the 'AllMotifEvo.RDS' file must be placed in IN_DATA folder.

#snrnaseq: Seurat object of snRNA-seq with annotation under the 'CellType' column.

#snatacseq: Seurat object of snATAC-seq with annotation under the 'CellType' column.

#plt_pref: Plot prefix

#outdir: Folder to output all data generated by this script

Rscript SCRIPTS/TFBS_Analyze.R \
			new_or_ms="MS" \
			snrnaseq="IN_DATA/snRNAseq.RDS" \
			snATACseq="IN_DATA/snATACseq.RDS" \
			plt_pref="Adult" \
			outdir='OUT_DATA/TFBS_Evo_Adult' \
```

<br>

### **RDG Analysis**
The following scripts identify all significant peak-gene linkage by using multiomic single-nucleus RNA-seq and single-nucleus ATAC-seq datasets. Then these outputs are used to identify genes that are linked to an excess amount of peaks divergent in a particular lineage.

**Step 1: Create peak-gene linkage statistics**
This script identifies all significant peak-gene linkage using Signac.

```
#fn_seurat: Seurat object for multiome data
#fn_gns: Protein coding genes divided into equal parts for parallelization. Must be a list in R in an RDS file
#index: The index of the gene list to run

Rscript SCRIPTS/Link_Peaks_To_Genes.R \
			fn_seurat='IN_DATA/multiome_seurat_object.RDS' \
			fn_gns='IN_DATA/pr_gns.RDS' \
			outdir='OUT_DATA/RDG' \
			index=1


# To parallelize across 5 threads
for i in {1..5}
do
Rscript SCRIPTS/Link_Peaks_To_Genes.R \
			fn_seurat='IN_DATA/multiome_seurat_object.RDS' \
			fn_gns='' \
			index=$i &
done
```

**Step 2: Call regulatorily divergent genes (RDGs)**
This script uses the peak-gene linkages identified in Step 1 to identify genes that are significantly more linked to peaks divergent in a particular lineage.

```

#new_or_ms: Should the file be the one in the manuscript (MS) or generated new using the previous command (NEW)? If MS is chosen, the 'Link_Peaks_To_Genes_Result.RDS' file must be placed in IN_DATA folder.

#nthreads: Number of threads for parallelization

#obs_peaksN: Minimum number of divergent peaks linked to the gene for it to be called an RDG.

#all_peaksN: Minimum number of all peaks linked to the gene for it to be called an RDG.

#fdr: FDR cutoff for RDG calls. Lower cutoffs indicate a greater difference between number of linked divergent peaks compared to randomized background

Rscript SCRIPTS/Call_RDGs.R \
			new_or_ms="MS" \
			nthreads=23 \
			obs_peaksN=2 \
			all_peaksN=5 \
			fdr=0.05 \
			plt_pref='Adult' \
			outdir='OUT_DATA/RDG' \
```






